lib/Sema/MiscDiagnostics.cpp

```diff
@@ -4657,6 +4657,62 @@ static void diagnoseComparisonWithNaN(const Expr *E, const DeclContext *DC) {
   const_cast<Expr *>(E)->walk(Walker);
 }
 
+static void diagUnqualifiedAccessToMethodNamedSelf(const Expr *E,
+                                                   const DeclContext *DC) {
+  if (!E || isa<ErrorExpr>(E) || !E->getType())
+    return;
+
+  class DiagnoseWalker : public ASTWalker {
+    ASTContext &Ctx;
+    const DeclContext *DC;
+
+  public:
+    DiagnoseWalker(const DeclContext *DC) : Ctx(DC->getASTContext()), DC(DC) {}
+
+    bool shouldWalkIntoSeparatelyCheckedClosure(ClosureExpr *expr) override {
+      return false;
+    }
+
+    bool shouldWalkIntoTapExpression() override { return false; }
+
+    std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+      if (!E || isa<ErrorExpr>(E) || !E->getType())
+        return {false, E};
+
+      if (auto *declRefExpr = dyn_cast<DeclRefExpr>(E)) {
+        if (declRefExpr->getDecl()->getBaseName() == Ctx.Id_self &&
+            declRefExpr->getType()->is<AnyFunctionType>()) {
+          if (auto typeContext = DC->getInnermostTypeContext()) {
+            auto parentCallExpr =
+                dyn_cast_or_null<CallExpr>(Parent.getAsExpr());
+            // self() is not easily confusable
+            if (!parentCallExpr || parentCallExpr->getFn() != declRefExpr) {
```

Comment:
Isn't the second condition always `false` here? `DeclRefExpr` cannot appear directly in argument list and if it was e.g. wrapped in parens then `Parent` wouldn't be a call.

Reply:
fixed, thanks! this was left over from an older implementation and missed that it's no longer needed